addons/microsoft_calendar/models/calendar_recurrence_rule.py

```diff
@@ -76,13 +70,79 @@ def _restart_microsoft_sync(self):
+        current_rrule = self.rrule
+        # event_tz is written on event in Microsoft but on recurrence in Odoo
+        vals['event_tz'] = microsoft_event.start.get('timeZone')
+        super()._write_from_microsoft(microsoft_event, vals)
+        base_event_time_fields = ['start', 'stop', 'allday']
+        new_event_values = self.env["calendar.event"]._microsoft_to_odoo_values(microsoft_event)
+        old_event_values = self.base_event_id and self.base_event_id.read(base_event_time_fields)[0]
+        if old_event_values and any(new_event_values[key] != old_event_values[key] for key in base_event_time_fields):
+            # we need to recreate the recurrence, time_fields were modified.
+            base_event_id = self.base_event_id
+            # We archive the old events to recompute the recurrence. These events are already deleted on Microsoft side.
+            # We can't call _cancel because events without user_id would not be deleted
+            (self.calendar_event_ids - base_event_id).microsoft_id = False
+            (self.calendar_event_ids - base_event_id).unlink()
+            base_event_id.with_context(dont_notify=True).write(dict(new_event_values, microsoft_id=False, need_sync_m=False))
+            if self.rrule == current_rrule:
+                # if the rrule has changed, it will be recalculated below
+                # There is no detached event now
+                self.with_context(dont_notify=True)._apply_recurrence()
+        else:
+            time_fields = (
+                    self.env["calendar.event"]._get_time_fields()
+                    | self.env["calendar.event"]._get_recurrent_fields()
+            )
+            # We avoid to write time_fields because they are not shared between events.
+            self._write_events(dict({
+                field: value
+                for field, value in new_event_values.items()
+                if field not in time_fields
+                }, need_sync_m=False)
+            )
+        # We apply the rrule check after the time_field check because the microsoft_id are generated according
+        # to base_event start datetime.
+        if self.rrule != current_rrule:
+            detached_events = self._apply_recurrence()
+            detached_events.microsoft_id = False
+            detached_events.unlink()
+
+    def _create_from_microsoft(self, microsoft_event, vals_list):
+        for gevent, vals in zip(microsoft_event, vals_list):
+            base_values = dict(
+                self.env['calendar.event']._microsoft_to_odoo_values(gevent),  # FIXME default reminders
+                need_sync_m=False,
+            )
+            # If we convert a single event into a recurrency on Microsoft, we should reuse this event on Odoo
+            # Microsoft reuse the event microsoft_id to identify the recurrence in that case
+            base_event = self.env['calendar.event'].search([('microsoft_id', '=', vals['microsoft_id'])])
```

Comment:
I would like to avoid the search in a for loop here. We have all the microsoft_ids from the vals_list.